### ISSUE TITLE
PVProperty for uint8 numpy array

### ISSUE THREAD

**Issue by vstadnytskyi** [caproto/caproto#704]
So I have found how to create a PV with PVProperty which in my case is an image encoded as a flat uint8 numpy vector (18432000 bytes long). The PV datatype can be selected via `dtype` property. Two dtype that seem to work fine are: float and int. Both of them inflate the size of the transmitted array. I have tried to use CHAR, but it somehow gets slowdown a lot. I suspect maybe there is some conversion happens on the caproto side.

Is there a way to specify the datatype to transfer uint8 numpy arrays? Or do I need to repackage my uint8 into int16 and use ChannelType.INT?

```
import numpy as np
raw_image = np.zeros((int(3000*4096*1.5),1), dtype = 'uint8')
image = pvproperty(value=raw_image.flatten(), dtype=caproto.ChannelType.INT, read_only=True)
```
I also found out what type are supported (https://nsls-ii.github.io/caproto/api.html#caproto.ChannelType) 

STRING = 0 (1 byte)
INT = 1 -> (int16)
FLOAT = 2 ->(flot64)
ENUM = 3
CHAR = 4
LONG = 5
DOUBLE = 6

**Comment by vstadnytskyi** [comment]
@klauer Thank you! 

Why would you assume that numpy is not installed? I cannot even imagine my life without numpy. I always have it everywhere as a prerequisite. For me, it would be fine to have it as a requirement for caproto.

**Comment by vstadnytskyi** [comment]
@klauer 

I have changed the dtype to bytes but now the server hangs. When I press Ctrl+C, it exists exactly where you pointed the potential bottleneck. When I change the type back to int, everything works.

> https://github.com/caproto/caproto/blob/10dc691852b9d4584cb2541a1ba09b6bf7b33325/caproto/_data.py#L980-L989


```
Traceback (most recent call last):
  File "camera_DLS.py", line 123, in <module>
    run(ioc.pvdb, **run_options)
  File "/home/femtoland/.local/lib/python3.8/site-packages/caproto/server/__init__.py", line 11, in run
    return run(pvdb, **kwargs)
  File "/home/femtoland/.local/lib/python3.8/site-packages/caproto/asyncio/server.py", line 267, in run
    loop.run_until_complete(task)
  File "/usr/lib/python3.8/asyncio/base_events.py", line 603, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.8/asyncio/base_events.py", line 570, in run_forever
    self._run_once()
  File "/usr/lib/python3.8/asyncio/base_events.py", line 1859, in _run_once
    handle._run()
  File "/usr/lib/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "/home/femtoland/.local/lib/python3.8/site-packages/caproto/asyncio/server.py", line 250, in start_server
    ret = await ctx.run(log_pv_names=log_pv_names)
  File "/home/femtoland/.local/lib/python3.8/site-packages/caproto/asyncio/server.py", line 210, in run
    await asyncio.gather(*tasks.tasks)
  File "/home/femtoland/.local/lib/python3.8/site-packages/caproto/asyncio/server.py", line 262, in run
    loop.run_until_complete(task)
  File "/usr/lib/python3.8/asyncio/base_events.py", line 603, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.8/asyncio/base_events.py", line 570, in run_forever
    self._run_once()
  File "/usr/lib/python3.8/asyncio/base_events.py", line 1859, in _run_once
    handle._run()
  File "/usr/lib/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "/home/femtoland/.local/lib/python3.8/site-packages/caproto/server/server.py", line 152, in _server_startup
    return await self.startup(self, async_lib)
  File "camera_DLS.py", line 76, in image
    await self.image.write(io_dict[key])
  File "/home/femtoland/.local/lib/python3.8/site-packages/caproto/_data.py", line 493, in write
    value = self.preprocess_value(value)
  File "/home/femtoland/.local/lib/python3.8/site-packages/caproto/_data.py", line 977, in preprocess_value
    value = b''.join(map(bytes, ([v] for v in value)))
  File "/home/femtoland/.local/lib/python3.8/site-packages/caproto/_data.py", line 977, in <genexpr>
    value = b''.join(map(bytes, ([v] for v in value)))
KeyboardInterrupt

```

**Comment by vstadnytskyi** [comment]
in the case of dtype = bytes I get the transfer of an image in 12 seconds and for dtype = int the image gets transferred in 0.2s.

/ I am uploading my smallest code right now, I have some issues with my local git I need to sort out./

**Comment by klauer** [comment]
> Why would you assume that numpy is not installed? I cannot even imagine my life without numpy. I always have it everywhere as a prerequisite. For me, it would be fine to have it as a requirement for caproto.

It sure would make things easier! But we pride ourselves on caproto having no "real" requirement other than Python itself.

I'd like to see this conversion issue resolved. We should be able to match - probably beat - the `dtype=int` timing as the data packet size would be halved.

**Comment by vstadnytskyi** [comment]
@klauer 

I have updated my comment above. Maybe I should have written a new one instead. anyway, yo sum

pv property dtype => resulting dtype of an array => time to read from a client
int =>  int32 => ~ 200 ms 
bytes => uint8 => 8 seconds
ChannelType.INT => int16 => ~100 ms - (probably to do with the server itself.)

**Comment by vstadnytskyi** [comment]
Few very important pieces I didn't mention are the caproto, python and OS versions.

BTW, the tests were done with the server and client on the same machine so no real network involved. 
**OS:**
MacOS Mojave 10.14.6
**Python version:** 
Python 3.7.4 (v3.7.4:e09359112e, Jul  8 2019, 14:54:52) 
**caproto version:** 
```
In [2]: caproto.__version__                                                                                                                                                                                                            
Out[2]: '0.4.3'
```

**Comment by klauer** [comment]
Initial fix should be in #705 with an example in the IOC section: 

https://github.com/caproto/caproto/blob/68bbba58d82bbf94ae273397592925ca7940ecf6/caproto/ioc_examples/big_image_noisy_neighbor.py#L24-L28

I'll close this for now, but feel free to reopen or comment as to whether or not the fix was sufficient.
